mev-boost-rs/src/relay.rs

```diff
@@ -0,0 +1,144 @@
+    }
+}
+
+#[derive(Clone)]
+pub struct Relay {
+    api: Client,
+    public_key: BlsPublicKey,
+}
+
+impl Relay {
+    pub fn api(&self) -> &Client {
+        &self.api
+    }
+
+    pub fn public_key(&self) -> &BlsPublicKey {
+        &self.public_key
+    }
+}
+
+impl From<RelayEndpoint> for Relay {
+    fn from(value: RelayEndpoint) -> Self {
+        let RelayEndpoint { url, public_key } = value;
+        Self { api: Client::new(BeaconClient::new(url)), public_key }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use ethereum_consensus::crypto::SecretKey;
+
+    use std::ops::Deref;
+
+    const URL: &str = "https://relay.com";
+
+    fn random_bls_public_key() -> BlsPublicKey {
+        let mut rng = rand::thread_rng();
+        let sk = SecretKey::random(&mut rng).unwrap();
+        sk.public_key()
+    }
+
+    #[test]
+    fn parse_relay_endpoint() {
+        let public_key = random_bls_public_key();
+        let public_key_hex = format!("{:#x}", public_key.deref());
+
+        let mut url = Url::parse(URL).unwrap();
+        url.set_username(&public_key_hex).unwrap();
+
+        let endpoint = RelayEndpoint::try_from(url.clone()).unwrap();
+        assert_eq!(endpoint.url, url);
+        assert_eq!(endpoint.public_key, public_key);
+    }
+
+    #[test]
+    fn parse_relay_endpoint_missing_public_key() {
+        let url = Url::parse(URL).unwrap();
+
+        let endpoint = RelayEndpoint::try_from(url);
+        assert!(std::matches!(endpoint, Err(RelayUrlError::MissingPublicKey(..))));
+    }
+
+    #[test]
+    fn parse_relay_endpoint_missing_0x_prefix() {
+        let public_key = random_bls_public_key();
+
+        // Format public key without '0x' prefix.
+        let public_key_hex = format!("{:x}", public_key.deref());
+
+        let mut url = Url::parse(URL).unwrap();
+        url.set_username(&public_key_hex).unwrap();
+
+        let endpoint = RelayEndpoint::try_from(url.clone());
+        assert!(std::matches!(endpoint, Err(RelayUrlError::Missing0xPrefix(..))));
+    }
+
+    #[test]
+    fn parse_relay_endpoint_invalid_hex() {
+        // Use string with proper '0x' prefix but invalid hex.
+        let invalid_hex = "0xethereum";
+
+        let mut url = Url::parse(URL).unwrap();
+        url.set_username(invalid_hex).unwrap();
+
+        let endpoint = RelayEndpoint::try_from(url.clone());
+        assert!(std::matches!(endpoint, Err(RelayUrlError::Hex(..))));
+    }
+
+    #[test]
+    fn parse_relay_endpoint_invalid_bls() {
+        let public_key = random_bls_public_key();
+
+        // Append some extra hex to the BLS public key.
+        let extra = "00";
+        let invalid_public_key_hex = format!("{:#x}{extra}", public_key.deref());
+
+        let mut url = Url::parse(URL).unwrap();
+        url.set_username(&invalid_public_key_hex).unwrap();
+
+        let endpoint = RelayEndpoint::try_from(url.clone());
+        assert!(std::matches!(endpoint, Err(RelayUrlError::Bls(..))));
```

Comment:
tests are nice! although we may be able to drop some of them depending on how much we lean on existing code already defined in other places

Reply:
I just kept the basic cases: the happy path and the case of a missing public key